### PR TITLE
[SHIRO-290] Implement BCrypt (and possibly other unix crypt formats like scrypt)

### PR DESCRIPTION
**Pull request by bmarwell**
- [X] Create AbstractCryptClass for Unix-Crypt-based algorithms (e.g. bcrypt, script, argon2)
- [X] Move some of the password matching logic into the Hash classes
    Reviwers: Review method placement and HAsh class description.
- [X] removed hashrequest
- [X] API change: made salt not-nullable.
    Additional constructor is supplied for hashing without or with
    default salt, the former and other methods/fields using
    SimpleByteSource.empty().
    Reviewers: Pay attention to method logic, so no empty salt is
    being used where a former `null` value would have created a
    new, random salt.
- [X] Modified tests to not expect exceptions in certain cases.
- [X] Modified tests to not expect passwordService calls when supplying an
    existing hash.
- [x] Fix Javadocs
- [x] Fix Hasher utility
- [x] Add Argon2 algorithms
- [x] Deprecate old non-KDF hash classes
- [ ] ~~Also print warnings when encountering the old hashes~~
- [x] What should `getBytes()` return? There is no overridden Javadoc. Currently, only the hashed data is returned, but maybe it was meant to also return the hash here? If not, why even implement `CodecSupport`? This needs to be clarified before merging.
   => This is more important if we want to retain Hex and Base64 formats. They need all parameters encoded.
 - [x] BCrypt iterations vs cost: What do we return for `getIterations()`? Currently I return the cost. But we could also return the real iterations (which is `2^cost`) it that made sense.
 - [X] ~~How to store argon2 parameters in Shiro1CryptFormat? Maybe create a Shiro2 CryptFormat?
   Argon2 outputs this: `$argon2i$v=19$m=65536,t=2,p=4$c29tZXNhbHQ$RdescudvJCsgt3ub+b+dWRWJTmaaJObG`.
  The Shiro1 crypt format only has: prefix(shiro1), algo, iterationcount, salt, hash.
  We would also need version and parameters or so, which could be empty for others.
  The current working copy uses a comma separated list for the iterationCount field, but I am not happy with that.~~
- [X] Shiro2CryptFormat (see comment above)?
- [X] revert Shiro1CryptFormat modifications?
- [X] ~~Can we drop Hex and Base64 Formats?~~ => Can be done in another PR.
- [X] Add more tests for bcrypt and argon2.
- [ ] ~~Tune Argon2 default parameters. Target computation time on servers: ~ 0.1s as per RFC recommendation.~~
- [ ] ~~Add scrypt. It is better than bcrypt.~~ => Another PR.
- [x] What does `DefaultPasswordService.testStringComparisonWhenNotUsingAParsableHashFormat` actually test? It only works if specific parameters are set.

## API Changes

- Changed default DefaultPasswordService.java algorithm to "Argon2id".
- `PasswordService.encryptPassword(Object plaintext)` will now throw
       a NullPointerException on `null` parameter. It was never specified how this method would behave.
- made salt not-nullable.

---

Following this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SHIRO) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SHIRO-XXX] - Fixes bug in SessionManager`,
       where you replace `SHIRO-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install apache-rat:check` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If you have a group of commits related to the same change, please squash your commits into one and force push your branch using `git rebase -i`. 

Trivial changes like typos do not require a JIRA issue (javadoc, comments...). 
In this case, just format the pull request title like `(DOC) - Add javadoc in SessionManager`.
 
If this is your first contribution, you have to read the [Contribution Guidelines](https://github.com/apache/shiro/blob/master/CONTRIBUTING.md)

If your pull request is about ~20 lines of code you don't need to sign an [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) 
if you are unsure please ask on the developers list.

To make clear that you license your contribution under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Review by fpapon — CHANGES_REQUESTED** [review]
@bmarwell we can add @since 2.0 to the new classes and the OpenBSD64 part come from another project so I think we have to mention it in the NOTICE file.

**Comment by bmarwell** [comment]
About `HexFormat` and `Base64Format`:

> `HexFormat.java`
> HashFormat that outputs only The hash's digest bytes in hex format. It does not print out anything else (salt, iterations, etc). This implementation is mostly provided as a convenience for command-line hashing.

and

> `Base64Format.java`
> HashFormat that outputs only the hash's digest bytes in Base64 format. It does not print out anything else (salt, iterations, etc). This implementation is mostly provided as a convenience for command-line hashing.

my +1 for dropping them:
1. for 2.0.0  mark them as deprecated and log warnings when being used.
2. for 2.1.0 throw exceptions when being used
3. for 2.2.0 complete removal

File: core/pom.xml [hunk 1]
```diff
@@ -101,6 +101,14 @@
             <groupId>org.apache.shiro</groupId>
             <artifactId>shiro-crypto-hash</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.shiro</groupId>
+            <artifactId>shiro-crypto-support-hashes-argon2</artifactId>
```
Comment (bdemers) [inline core/pom.xml:106]: nit: he groupId could be `org.apache.shiro.hashes` and artifactId: `shiro-crypto-argon2`
Or, `org.apache.shiro.crypto:shiro-hashes-argon2`

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/ConfigurableHashService.java [hunk 2]
```diff
@@ -18,44 +18,20 @@
...
-
     /**
      * Sets the name of the {@link java.security.MessageDigest MessageDigest} algorithm that will be used to compute
      * hashes.
      *
      * @param name the name of the {@link java.security.MessageDigest MessageDigest} algorithm that will be used to
      *             compute hashes.
      */
-    void setHashAlgorithmName(String name);
+    void setDefaultAlgorithmName(String name);
```
Comment (bdemers) [inline crypto/hash/src/main/java/org/apache/shiro/crypto/hash/ConfigurableHashService.java:35]: nit: might need to tweak the javadoc to go with this, I'm assuming this is the alg used if the hash cannot be determined?

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/ConfigurableHashService.java [hunk 3]
```diff
@@ -18,44 +18,20 @@
...
 
-    /**
-     * Sets a source of randomness used to generate public salts that will in turn be used during hash computation.
-     *
-     * @param rng a source of randomness used to generate public salts that will in turn be used during hash computation.
-     */
-    void setRandomNumberGenerator(RandomNumberGenerator rng);
```
Comment (bdemers) [inline crypto/hash/src/main/java/org/apache/shiro/crypto/hash/ConfigurableHashService.java:60]: +1 for removing this from the public API.

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/DefaultHashService.java [hunk 4]
```diff
@@ -123,222 +71,52 @@ public DefaultHashService() {
...
-     * to protect the integrity of the private salt (without a public salt, the private salt would be exposed as-is,
-     * which is undesirable).
-     *
-     * @param generatePublicSalt whether or not a public salt should be randomly generated and used to compute a hash
-     *                           if a {@link HashRequest} does not specify a salt.
-     */
-    public void setGeneratePublicSalt(boolean generatePublicSalt) {
-        this.generatePublicSalt = generatePublicSalt;
+    public void setRandom(Random random) {
+        this.random = random;
```
Comment (bdemers) [inline crypto/hash/src/main/java/org/apache/shiro/crypto/hash/DefaultHashService.java:120]: If possible make the random getter/setter package private (I'm guessing it's only used for testing?)

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/HashProvider.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        ServiceLoader<HashSpi<? extends Hash>> hashSpis = load();
+
+        return StreamSupport.stream(hashSpis.spliterator(), false)
+                .filter(hashSpi -> hashSpi.getImplementedAlgorithms().contains(algorithmName))
+                .findAny();
+    }
+
+    @SuppressWarnings("unchecked")
+    private static ServiceLoader<HashSpi<? extends Hash>> load() {
+        return (ServiceLoader<HashSpi<? extends Hash>>) (Object) ServiceLoader.load(HashSpi.class);
```
Comment (bdemers) [inline crypto/hash/src/main/java/org/apache/shiro/crypto/hash/HashProvider.java:57]: In JJWT, we ran into some issues with class loaders on various environments, IIRC depending on the env: vanilla java, JPMS, spring, OSGi, the actual classloader used could be different, so we ended up doing something like this:
https://github.com/jwtk/jjwt/blob/master/impl/src/main/java/io/jsonwebtoken/impl/lang/Services.java

(basically checking the three common classloaders)
I don't think this is blocking, as it shouldn't change the API of this class, but it's something to think about

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/Md2Hash.java [hunk 6]
```diff
@@ -30,10 +30,13 @@
  * techniques and how the overloaded constructors function.
  *
  * @since 0.9
+ * @deprecated and will throw exceptions since 2.0.0, to be removed in 2.1.0.
```
Comment (bdemers) [inline crypto/hash/src/main/java/org/apache/shiro/crypto/hash/Md2Hash.java:33]: Rip anything out in 2.0, removing anything is technically a breaking change. (it's easy to add back in later if/when needed)

Comment (bdemers) [reply crypto/hash/src/main/java/org/apache/shiro/crypto/hash/Md2Hash.java:33]: we can also mark them deprecated now, and then remove all the deprecated bits before the official 2.0 release if you are on the fence about anything.

Another idea, is we could nag users: print a log message: "WARN: Seriously, stop using MD2, here is why <insert some URL here>" 👺

File: crypto/support/hashes/bcrypt/src/main/java/org/apache/shiro/crypto/support/hashes/bcrypt/OpenBSDBase64.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+     *  distributed under the License is distributed on an "AS IS" BASIS,
+     *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     *  See the License for the specific language governing permissions and
+     *  limitations under the License.
+     */
+
+    /**
+     * A mod of Square's Okio Base64 encoder
+     * <p>
+     * Original author: Alexander Y. Kleymenov
```
Comment (bdemers) [inline crypto/support/hashes/bcrypt/src/main/java/org/apache/shiro/crypto/support/hashes/bcrypt/OpenBSDBase64.java:74]: If not already make sure this gets added to the NOTICE for this module

File: crypto/support/pom.xml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        <version>2.0.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>support</artifactId>
+    <name>Apache Shiro :: Cryptography :: Support</name>
+    <packaging>pom</packaging>
+
+    <modules>
+        <module>hashes</module>
```
Comment (bdemers) [inline crypto/support/pom.xml:38]: nit: you can remove the extra empty parent poms by doing something like:
```xml
<modules>
  <module>hashes/bcrypt</module>
  <module>hashes/argon2</module>
</modules>
```

File: tools/hasher/pom.xml [hunk 9]
```diff
@@ -50,7 +50,11 @@
         </dependency>
         <dependency>
             <groupId>org.slf4j</groupId>
-            <artifactId>slf4j-simple</artifactId>
+            <artifactId>slf4j-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
```
Comment (bdemers) [inline tools/hasher/pom.xml:57]: nice cleanup 😄

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/ConfigurableHashService.java (hunk 2 above)
Comment (bmarwell) [reply crypto/hash/src/main/java/org/apache/shiro/crypto/hash/ConfigurableHashService.java:35]: Yes, the javadocs are the unchecked checkbox ;) Next commit is on the way.

The default is being used when a `HashRequest` does not carry an algorithm name. Its old behaviour/semantics was at least similar, but I was not looking into making it a 100% compatible here.

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/DefaultHashService.java (hunk 4 above)
Comment (bmarwell) [reply crypto/hash/src/main/java/org/apache/shiro/crypto/hash/DefaultHashService.java:120]: Not even for testing. Removing it until we need it.

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/Md2Hash.java (hunk 6 above)
Comment (bmarwell) [reply crypto/hash/src/main/java/org/apache/shiro/crypto/hash/Md2Hash.java:33]: +1 for ripping out. We should nag the users in 1.7+ then.
And not just for MD2. For all iteration+salt+pepper based hashes, even if there is no alternative in shiro 1. Seriously.

File: tools/hasher/pom.xml (hunk 9 above)
Comment (bmarwell) [reply tools/hasher/pom.xml:57]: It is not so funny actually. See the chat in slack. To be able test things written to `System.out`, we need `logback-classic`, as the `simple-logger` cannot be intercepted.
Romain suggested to refactor the hasher utility a bit further, and I would like to do that in another PR.

File: crypto/hash/src/main/java/org/apache/shiro/crypto/hash/HashProvider.java (hunk 5 above)
Comment (bmarwell) [reply crypto/hash/src/main/java/org/apache/shiro/crypto/hash/HashProvider.java:59]: What bugs me most is the ugly casting to object and back. But with what I saw from your `jjwt` example, adding a parameter might help with that!

Comment (bmarwell) [reply crypto/hash/src/main/java/org/apache/shiro/crypto/hash/HashProvider.java]: I removed the generic type.

**Pull request closed without merging**